### PR TITLE
fixed regexes to avoid ReDoS attacks

### PR DESCRIPTION
**Pull request by jeffrey-pinyan-ithreat**
Replaces `[\s\S]*?` with less explosive alternatives.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -34,24 +19,24 @@ win32.parse = function(pathString) {
...
 };
 
 
 
-// Split a filename into [root, dir, basename, ext], unix version
+// Split a filename into [dir, root, basename, name, ext], unix version
 // 'root' is just a slash, or nothing.
 var splitPathRe =
-    /^(\/?|)([\s\S]*?)((?:\.{1,2}|[^\/]+?|)(\.[^.\/]*|))(?:[\/]*)$/;
+    /^((\/?)(?:[^\/]*\/)*)((\.{1,2}|[^\/]+?|)(\.[^.\/]*|))[\/]*$/;
```
Comment (wickedest) [inline index.js:39]: These look like breaking changes to me, and I am not entirely sure if this fixes the DoS.  I did a little work on this issue.  It is tempting to fix the groups, the whitespace (\s\S), etc., but I found that the existing unit tests were incomplete (even if some are unrealistic).

I found that these tests needed to be added:
```js
var regexDosFileNameLength = 50000;
var regexDosFileName = Buffer.alloc(regexDosFileNameLength, '/').toString() + 'file.txt';

[{ root: '/', dir: '/home/user/dir', base: 'file.txt', ext: '.txt', name: 'file' }, '/home/user/dir/file.txt/'],
[{ root: '/', dir: '/\n\thome/user/dir', base: 'file.txt', ext: '.txt', name: 'file' }, '/\n\thome/user/dir/file.txt/'],
[{ root: '/', dir: '/////////', base: 'file.txt', ext: '.txt', name: 'file' }, '//////////file.txt'],
[{ root: '/', dir: '/', base: '', ext: '', name: '' }, '//////////'],
[{ root: '/', dir: Buffer.alloc(regexDosFileNameLength - 1, '/').toString(), base: 'file.txt', ext: '.txt', name: 'file' }, regexDosFileName],
 ];
```

If you want these to be breaking changes, then fine.  The `regexDosFileNameLength` is specifically to test the regex DoS.

Comment (jeffrey-pinyan-ithreat) [reply index.js:39]: I can pass those tests with
```
var splitPathRe =
  /^((\/?)(?:[^\/]*\/)*)((\.{1,2}|[^\/]+?|)(\.[^.\/]*|))(?:(?<=^\/)|(?<!\/))[\/]*$/;`

return {
  // ...
  dir: allParts[0] === '/' ? allparts[0] : allParts[0].slice(0, -1),
  // ...
}```

But as thrilling as it is to have a single regex handling the parsing, it'd be much wiser to clean up the parameter (e.g. by stripping trailing slashes, while retaining a leading slash) and keeping the regex simple.

**Review by mwjsosgge2765252 — COMMENTED** [review]
inl_ctx_stk.pop_paired

**Review by immense055 — COMMENTED** [review]
.npmignore

**Review by immense055 — COMMENTED** [review]
.npmignore

**Comment by immense055** [comment]
09e1086512bd50f2767b8c32fa74c0ff0be4c8cd

**Review by JamesonDiamond1 — COMMENTED** [review]
@JamesonDiamond1

**Review by JamesonDiamond1 — COMMENTED** [review]
@JamesonDiamond1

**Pull request merged**